### ISSUE TITLE
Does Masking with ctc_cost_precise work ?

### ISSUE THREAD

**Issue by unrealwill** [daweileng/keras_MOD#5]
#model has been compiled with
model.compile(loss="ctc_cost_precise", optimizer="Adadelta", sample_weight_mode='temporal')

XMask = np.ones((X.shape[0],X.shape[1]),dtype=X.dtype )
XMask[0,-1] = 0.0   # it's fine without this line
res = model.train_on_batch(x=X, y=Y, sm_mask=XMask, return_sm=True)
print(res[0]) # -> inf or nan as soon as there is a 0.0

Can you please advise ?
Thank you

- [X] If running on Theano, check that you are up-to-date with the master branch of Theano. You can update with:
pip install git+git://github.com/Theano/Theano.git --upgrade --no-deps

**Comment by daweileng** [comment]
For this open-sourced implementation, I'd advise to run in `'float64'` mode when `ctc_cost_precise` is chosen as loss function. `float32` is not sufficient for the 'precise' version, after all it's meant to be precise. For NN training in `float32` mode, 'ctc_cost_for_train' would be preferred.

**Comment by unrealwill** [comment]
Thank you for the advice. I have set float64 in ~/.keras/keras.json and it does solve a lot of the numerical stability issues I was encountering.

But it's still not resolving the Masking issue, which is that as soon as there is a 0.0 in the input mask, return nan or inf at the first iteration.

Currently I'm using https://github.com/mcf06/theano_ctc (a wrapper of Baidu warp-ctc) (see tests folder for a keras example usage), it's (a lot) faster, and as far as I've tested on toy projects it's quite numerically stable even with keras in float32.

How does PreciseCTC compare to Baidu warp-ctc ? Was PreciseCTC superseded by warp-ctc ?

**Comment by daweileng** [comment]
Will Nan or Inf occur if `ctc_cost_for_train` is used? The masking mechanism is the same for `ctc_cost_for_train` and `ctc_cost_precise`.
About runtime speed, it's no wonder that warp-ctc is faster, since it's written in highly optimized C++ code. But be aware that their mask works a little differently.

**Comment by unrealwill** [comment]
Yes I don't get Nan with ctc_cost_for_train (But I don't get convergence on ~100 timesteps long problems) But I do get Nan with ctc_cost_precise (Which converge better but I can't seem to get masking to work).

**Comment by daweileng** [comment]
Strange. Could you provide your code and the full error trace back message?

**Comment by unrealwill** [comment]
Sure, here is the toy project I use to compare CTC : 
(Basically we take a sequence of integers we add a random number of zeros between each and we try to learn to drop the zeros)

No errors just printing nan
https://gist.github.com/unrealwill/4bd198c1a4b129837395b11d555252e2
